Chore: Query oauth info from a new instance

linoman opened the pull request:
**What is this feature?**

This small PR avoids a settings race condition between the Google OAuth configuration settings and the SSO settings.

**Why do we need this feature?**

To avoid a race condition.

**Who is this feature for?**

IAM

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

# Release notice breaking change

We're adding a validation between the response of the ID token HD parameter and the list of allowed domains as an extra layer of security. In the event that the HD parameter doesn't match the list of allowed domains, we're denying access to Grafana.

If you set Google OAuth configuration using `api_url,` you might be using the legacy implementation of OAuth, which doesn't have the HD parameter describing the organisation the approved token comes from. This could break your login flow.

This feature can be turned off through the configuration toggle `validate_hd `. Anyone using the legacy Google OAuth configuration should turn off this validation if the ID Token response doesn't have the HD parameter.

linoman committed:
Query oauth info from a new instance

linoman commented:
Upon implementation and review with @Jguer, the scope of this PR will be extended to include a configuration option to disable HD validation

linoman committed:
add hd validation flag

linoman committed:
add `disable_hd_validation` to settigns map

mgyongyosi commented on pkg/login/social/social.go:
```diff
@@ -63,6 +63,7 @@ type OAuthInfo struct {
 	AutoLogin               bool              `mapstructure:"auto_login" toml:"auto_login"`
 	ClientId                string            `mapstructure:"client_id" toml:"client_id"`
 	ClientSecret            string            `mapstructure:"client_secret" toml:"-"`
+	DisableHDValidation     bool              `mapstructure:"disable_hd_validation" toml:"disable_hd_validation"`
```
Please don't put this into the OAuthInfo if it only applies to the Google integration. Rather please create a custom key like: https://github.com/grafana/grafana/blob/3f2eb8bd6a57d1449ae7e6a76a98e69a58d5e8d5/pkg/login/social/connectors/azuread_oauth.go#L82

linoman replied:
I've moved the `DisableHDValidation` to the connector as requested. PTAL

linoman commented on pkg/login/social/connectors/google_oauth.go:
```diff
@@ -298,11 +300,17 @@ func (s *SocialGoogle) getGroupsPage(ctx context.Context, client *http.Client, u
-	if len(s.info.AllowedDomains) == 0 {
+	info := s.GetOAuthInfo()
+
+	if info.DisableHDValidation {
```
Yes! This toggle will become useful for instances where legacy code doesn't return the `hd` parameter; which it is still used by a lot of Grafana users.

This toggle was suggested by @Jguer in this [thread](https://raintank-corp.slack.com/archives/C05PJJ5JFAP/p1708608596916029?thread_ts=1708594382.367209&cid=C05PJJ5JFAP)

Jguer replied:
Unfortunately yes, if the account does not belong to a google workspace it will not return an hd

linoman committed:
address pr comments

linoman commented on pkg/login/social/connectors/google_oauth.go:
```diff
@@ -298,11 +300,17 @@ func (s *SocialGoogle) getGroupsPage(ctx context.Context, client *http.Client, u
 }
 
 func (s *SocialGoogle) isHDAllowed(hd string) error {
```
I've changed the signature as requested

linoman committed:
Merge remote-tracking branch 'origin/main' into linoman/avoid_race_condition_for_allowed_domain_validations

linoman committed:
fix tests

linoman committed:
fix tests

linoman committed:
address pr comments

mgyongyosi reviewed (approved):
Looks good to me. Before you merge it please add a test to verify that the extra field is fetched by `OAuthStrategy`: https://github.com/grafana/grafana/blob/main/pkg/services/ssosettings/strategies/oauth_strategy_test.go#L112

Jguer commented on pkg/login/social/connectors/google_oauth.go:
```diff
@@ -24,13 +24,16 @@ const (
 
 type SocialGoogle struct {
 	*SocialBase
+	disableHDValidation bool
```
avoid negative options

rename to validateHD bool will make it more readable

linoman replied:
I have updated the variable name as requested.

Jguer commented on pkg/login/social/connectors/google_oauth.go:
```diff
@@ -24,13 +24,16 @@ const (
 	legacyAPIURL            = "https://www.googleapis.com/oauth2/v1/userinfo"
 	googleIAMGroupsEndpoint = "https://content-cloudidentity.googleapis.com/v1/groups/-/memberships:searchDirectGroups"
 	googleIAMScope          = "https://www.googleapis.com/auth/cloud-identity.groups.readonly"
+	disableHDValidationKey  = "disable_hd_validation"
```
proposed change:

`validate_hd`

linoman replied:
I have updated the configuration option name as requested.

linoman committed:
address pr comments and replace strings for provider names

linoman committed:
Merge remote-tracking branch 'origin/main' into linoman/avoid_race_condition_for_allowed_domain_validations

Jguer reviewed (changes requested):
avoid negative options

linoman committed:
rename configuration option

linoman committed:
update documentation

linoman committed:
update documentation

linoman commented:
@mgyongyosi Thank you for your comments. I replaced the provider's name in the `oauth_strategy_test` for the constants available in `social`.

@Jguer, I have replaced the names of the variables/conf options and added some descriptions in the docs. Additionally, I've added the new configuration option to the `default.ini` and `sample.ini` files.

Jguer commented on conf/defaults.ini:
```diff
@@ -679,6 +679,7 @@ token_url = https://oauth2.googleapis.com/token
 api_url = https://openidconnect.googleapis.com/v1/userinfo
 signout_redirect_url =
 allowed_domains =
+validate_hd =
```
```suggestion
validate_hd = true
```

Jguer commented on pkg/login/social/connectors/google_oauth.go:
```diff
@@ -45,6 +48,7 @@ type googleUserData struct {
 func NewGoogleProvider(info *social.OAuthInfo, cfg *setting.Cfg, ssoSettings ssosettings.Service, features featuremgmt.FeatureToggles) *SocialGoogle {
 	provider := &SocialGoogle{
 		SocialBase: newSocialBase(social.GoogleProviderName, info, features, cfg),
+		validateHD: MustBool(info.Extra[validateHDKey], false),
```
setting it by default to false?

linoman replied:
Changed as requested

Jguer commented on docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:
```diff
@@ -111,6 +111,12 @@ automatically signed up.
 OAuth 2.0 authentication API. Refer to Google's OAuth [documentation](https://developers.google.com/identity/openid-connect/openid-connect#hd-param).
 
+{{% admonition type="note" %}}
+Since Grafana 10.3.0, the `hd` parameter retrieved from Google ID token is used to determine the user's hosted domain. The Google Oauth `allowed_domains` configuration option is used to restrict access to users from a specific domain. If the `allowed_domains` configuration option is set, the `hd` parameter from the Google ID token must match the `allowed_domains` configuration option. If the `hd` ...
```
```suggestion
Since Grafana 10.3.0, the `hd` parameter retrieved from Google ID token is also used to determine the user's hosted domain. The Google Oauth `allowed_domains` configuration option is used to restrict access to users from a specific domain. If the `allowed_domains` configuration option is set, the `hd` parameter from the Google ID token must match the `allowed_domains` configuration option. If the `hd` parameter from the Google ID token does not match the `allowed_domains` configuration option, the user is denied access.
```

Jguer commented on docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:
```diff
@@ -111,6 +111,12 @@ automatically signed up.
+{{% admonition type="note" %}}
+Since Grafana 10.3.0, the `hd` parameter retrieved from Google ID token is used to determine the user's hosted domain. The Google Oauth `allowed_domains` configuration option is used to restrict access to users from a specific domain. If the `allowed_domains` configuration option is set, the `hd` parameter from the Google ID token must match the `allowed_domains` configuration option. If the `hd` ...
+
+This validation is enabled by default. To disable this validation, set the `validate_hd` configuration option to `false`.
```
```suggestion
When an account does not belong to a google workspace, the hd claim will not be available.

This validation is enabled by default. To disable this validation, set the `validate_hd` configuration option to `false`. The `allowed_domains` configuration option will use the email claim to validate the domain.
```

linoman committed:
Apply suggestions from code review

Co-authored-by: Jo <joao.guerreiro@grafana.com>

linoman committed:
address PR comments

linoman commented:
Backports
- [10.3.x](https://github.com/grafana/grafana/pull/83725)
- [10.4.x](https://github.com/grafana/grafana/pull/83726)

mjseaman commented:
@linoman I think there's a word missing in your breaking change description:

> We're adding a [WORD MISSING] between the response of the ID token HD parameter and the list of allowed domains...
